docs/rfc/active/5818-serial-updates.md

```diff
@@ -0,0 +1,380 @@
+
+### Change Processor
+
+Check the allowed set of transitions between current and proposed aspect state. This would be achieved with
+business-specific logic injected into the DataHub codebase via plugins. This feels like the wrong approach.
+
+Good
+* Fast as all code is executed in GMS
+
+Bad
+* No guarantee of state between the read and write
+* No guarantee that the plugin code will run
+* Specific business logic getting tied up with GMS code
+
+```mermaid
+sequenceDiagram
+    actor Client
+    Client->>+GMS: IngestAspect
+    GMS->>MetadataStore: GetAspect
+    MetadataStore-->>GMS: aspect
+    GMS->>+ChangeProcessor: IsPermitted?
+    alt permitted
+        ChangeProcessor-->>GMS: yes
+        note right of MetadataStore: Aspect may have changed here
+        GMS->>MetadataStore: store new version
+        MetadataStore-->>GMS: result
+    else disallowed
+        ChangeProcessor-->>-GMS: no
+    end
+    GMS-->>-Client: result
+```
+
+### Idempotent Partial Update (Patch)
+
+The client proposes a partial change to an aspect that is idempotent. The change can apply whatever state the aspect
+is in at the time. This will need to be extended to check for multiple currentAspect states when requested.
+
+Good
+* No need for business-specific logic in GMS code
+* Fast, as retry loop will be confined to GMS
+
+Bad
+* GMS will need to understand partial (PATCH) updates
+* Retries will need to be handled in GMS
+
+```mermaid
+sequenceDiagram
+    actor Client
+    Client->>GMS: UpdateAspect
+    loop Attempt until applied or error
+    GMS->>MetadataStore: GetAspect
+    MetadataStore-->>GMS: currentAspect
+    GMS-->>GMS: Locally modify aspect
+    GMS->>MetadataStore: UpdateAspect if still currentAspect
+    MetadataStore-->>GMS: applied?
+    end
+    alt applied
+        GMS-->>Client: OK
+    else error
+        GMS-->>Client: ERROR
+    end
+```
+
+#### Patch Language
+
+If a patch language were available, it might be possible to update collections in such a way that it's not
+necessary for clients to know about the previous state of an aspect.
+
+MongoDB offers an update mode which allows clients to [add items to a set](https://mongodb.github.io/mongo-java-driver/4.7/apidocs/mongodb-driver-core/com/mongodb/client/model/Updates.html#addToSet(java.lang.String,TItem)),
+for example. This would avoid the need to go back to the client to ask them to construct the final state of a
+collection.
+
+##### Implementation of PATCH with json-patch format
+
+Client sends its update request in json-patch or some other delta format. GMS will reject the update if the "previous"
+side of the delta does not match. This closely matches the work in https://github.com/datahub-project/datahub/pull/5901.
+
+### Serialised Updates Exclusively via Kafka
+
+Here, the client never updates the aspect via HTTP but instead passes it via the MetadataChangeProposal topic in Kafka.
```

Comment:
I think it's possible that we do provide an optional version tag in the Kafka ingest path as well, but there is a higher likelihood of out of date updates. Also, it's obviously not possible for a client to perform a retry since they won't know what failed...